File: csharp/src/Json.cs
// Just enough JSON for sekreto.
//
// System.Text.Json is part of the base class library, not a third-party
// package, so this is a thin adapter onto it rather than a parser of its
// own: JSON in, plain Dictionary/List/string/double/bool/null out - and
// back again for the CLI's line of output.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace Voxgig.Sekreto
{
    public static class Json
    {
        /// <summary>Parse JSON text, answering null for anything unreadable.</summary>
        public static object Parse(string text)
        {
            return TryParse(text, out object value) ? value : null;
        }

        /// <summary>
        /// Parse JSON text, answering false when the text is not JSON at all
        /// - which a null answer from Parse cannot say, since JSON `null`
        /// parses to the same thing.
        /// </summary>
        internal static bool TryParse(string text, out object value)
        {
            value = null;

            if (string.IsNullOrEmpty(text))
            {
                return false;
            }

            try
            {
                using var doc = JsonDocument.Parse(text);
                value = Convert(doc.RootElement);
                return true;
            }
            catch (JsonException)
            {
                return false;
            }
        }

        private static object Convert(JsonElement element)
        {
            switch (element.ValueKind)
            {
                case JsonValueKind.Object:
                    var map = new Dictionary<string, object>();
                    foreach (JsonProperty entry in element.EnumerateObject())
                    {
                        map[entry.Name] = Convert(entry.Value);
                    }
                    return map;

                case JsonValueKind.Array:
                    var list = new List<object>();
                    foreach (JsonElement entry in element.EnumerateArray())
                    {
                        list.Add(Convert(entry));
                    }
                    return list;

                case JsonValueKind.String:
                    return element.GetString();

                case JsonValueKind.Number:
                    return element.GetDouble();

                case JsonValueKind.True:
                    return true;

                case JsonValueKind.False:
                    return false;

                default:
                    return null;
            }
        }

        /// <summary>Render a value as compact JSON.</summary>
        public static string Stringify(object value)
        {
            var out_ = new StringBuilder();
            Write(value, out_);
            return out_.ToString();
        }

        private static void Write(object value, StringBuilder out_)
        {
            switch (value)
            {
                case null:
                    out_.Append("null");
                    return;

                case string text:
                    Quote(text, out_);
                    return;

                case bool flag:
                    out_.Append(flag ? "true" : "false");
                    return;

                case double number:
                    // 5.0 prints as 5, matching every other port.
                    out_.Append(number == Math.Truncate(number) && Math.Abs(number) < 1e15
                        ? ((long)number).ToString(CultureInfo.InvariantCulture)
                        : number.ToString("R", CultureInfo.InvariantCulture));
                    return;

                case int number:
                    out_.Append(number.ToString(CultureInfo.InvariantCulture));
                    return;

                case Dictionary<string, object> map:
                    out_.Append('{');
                    bool firstentry = true;
                    foreach (var entry in map)
                    {
                        if (!firstentry)
                        {
                            out_.Append(',');
                        }
                        firstentry = false;
                        Quote(entry.Key, out_);
                        out_.Append(':');
                        Write(entry.Value, out_);
                    }
                    out_.Append('}');
                    return;

                case List<object> list:
                    out_.Append('[');
                    bool firstitem = true;
                    foreach (object item in list)
                    {
                        if (!firstitem)
                        {
                            out_.Append(',');
                        }
                        firstitem = false;
                        Write(item, out_);
                    }
                    out_.Append(']');
                    return;

                default:
                    Quote(System.Convert.ToString(value, CultureInfo.InvariantCulture), out_);
                    return;
            }
        }

        private static void Quote(string text, StringBuilder out_)
        {
            out_.Append('"');

            foreach (char head in text)
            {
                switch (head)
                {
                    case '"': out_.Append("\\\""); break;
                    case '\\': out_.Append("\\\\"); break;
                    case '\n': out_.Append("\\n"); break;
                    case '\r': out_.Append("\\r"); break;
                    case '\t': out_.Append("\\t"); break;
                    default:
                        if (0x20 > head)
                        {
                            out_.Append("\\u").Append(((int)head).ToString("x4"));
                        }
                        else
                        {
                            out_.Append(head);
                        }
                        break;
                }
            }

            out_.Append('"');
        }
    }
}

File: rust/src/json.rs
//! Minimal JSON support for sekreto.
//!
//! sekreto adds no third-party dependencies, so it carries just enough JSON
//! to read a vault's answer and write the CLI's own line of output. It is
//! deliberately not a general-purpose library.

use std::collections::BTreeMap;

/// A JSON value.
#[derive(Clone, Debug, PartialEq)]
pub enum Json {
    Null,
    Bool(bool),
    Num(f64),
    Str(String),
    List(Vec<Json>),
    Map(BTreeMap<String, Json>),
}

impl Json {
    /// The named entry of a map, if this is a map that has it.
    pub fn get(&self, key: &str) -> Option<&Json> {
        match self {
            Json::Map(entries) => entries.get(key),
            _ => None,
        }
    }

    /// This value as a string, the way the canonical port's String() renders
    /// it: strings verbatim, scalars as text, containers as JSON.
    pub fn text(&self) -> String {
        match self {
            Json::Null => "null".to_string(),
            Json::Bool(value) => value.to_string(),
            Json::Num(value) => numstr(*value),
            Json::Str(value) => value.clone(),
            _ => stringify(self),
        }
    }
}

/// Render a number without a trailing `.0`, so 5.0 prints as `5`.
pub fn numstr(value: f64) -> String {
    if value.is_finite() && value == value.trunc() && value.abs() < 1e15 {
        return format!("{}", value as i64);
    }
    format!("{}", value)
}

/// Render a value as compact JSON.
pub fn stringify(value: &Json) -> String {
    match value {
        Json::Null => "null".to_string(),
        Json::Bool(entry) => entry.to_string(),
        Json::Num(entry) => numstr(*entry),
        Json::Str(entry) => quote(entry),
        Json::List(entries) => {
            let parts: Vec<String> = entries.iter().map(stringify).collect();
            format!("[{}]", parts.join(","))
        }
        Json::Map(entries) => {
            let parts: Vec<String> = entries
                .iter()
                .map(|(key, entry)| format!("{}:{}", quote(key), stringify(entry)))
                .collect();
            format!("{{{}}}", parts.join(","))
        }
    }
}

/// Quote a string as JSON.
pub fn quote(text: &str) -> String {
    let mut out = String::from("\"");

    for head in text.chars() {
        match head {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            _ if (head as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", head as u32)),
            _ => out.push(head),
        }
    }

    out.push('"');
    out
}

/// Parse JSON text, answering None for anything unreadable.
pub fn parse(text: &str) -> Option<Json> {
    let chars: Vec<char> = text.chars().collect();
    let mut at = 0usize;
    let value = value(&chars, &mut at)?;
    Some(value)
}

fn skip(chars: &[char], at: &mut usize) {
    while *at < chars.len() && chars[*at].is_whitespace() {
        *at += 1;
    }
}

fn value(chars: &[char], at: &mut usize) -> Option<Json> {
    skip(chars, at);

    if *at >= chars.len() {
        return None;
    }

    match chars[*at] {
        '{' => map(chars, at),
        '[' => list(chars, at),
        '"' => string(chars, at).map(Json::Str),
        't' => literal(chars, at, "true", Json::Bool(true)),
        'f' => literal(chars, at, "false", Json::Bool(false)),
        'n' => literal(chars, at, "null", Json::Null),
        _ => number(chars, at),
    }
}

fn literal(chars: &[char], at: &mut usize, word: &str, value: Json) -> Option<Json> {
    for (offset, head) in word.chars().enumerate() {
        if *at + offset >= chars.len() || chars[*at + offset] != head {
            return None;
        }
    }
    *at += word.len();
    Some(value)
}

fn map(chars: &[char], at: &mut usize) -> Option<Json> {
    let mut out = BTreeMap::new();
    *at += 1; // {
    skip(chars, at);

    if *at < chars.len() && '}' == chars[*at] {
        *at += 1;
        return Some(Json::Map(out));
    }

    loop {
        skip(chars, at);
        let key = string(chars, at)?;
        skip(chars, at);

        if *at >= chars.len() || ':' != chars[*at] {
            return None;
        }
        *at += 1;

        out.insert(key, value(chars, at)?);
        skip(chars, at);

        if *at >= chars.len() {
            return None;
        }

        if ',' == chars[*at] {
            *at += 1;
            continue;
        }

        *at += 1; // }
        break;
    }

    Some(Json::Map(out))
}

fn list(chars: &[char], at: &mut usize) -> Option<Json> {
    let mut out = Vec::new();
    *at += 1; // [
    skip(chars, at);

    if *at < chars.len() && ']' == chars[*at] {
        *at += 1;
        return Some(Json::List(out));
    }

    loop {
        out.push(value(chars, at)?);
        skip(chars, at);

        if *at >= chars.len() {
            return None;
        }

        if ',' == chars[*at] {
            *at += 1;
            continue;
        }

        *at += 1; // ]
        break;
    }

    Some(Json::List(out))
}

fn string(chars: &[char], at: &mut usize) -> Option<String> {
    if *at >= chars.len() || '"' != chars[*at] {
        return None;
    }

    let mut out = String::new();
    *at += 1;

    while *at < chars.len() {
        let head = chars[*at];
        *at += 1;

        if '"' == head {
            return Some(out);
        }

        if '\\' != head {
            out.push(head);
            continue;
        }

        if *at >= chars.len() {
            return None;
        }

        let next = chars[*at];
        *at += 1;

        match next {
            'n' => out.push('\n'),
            'r' => out.push('\r'),
            't' => out.push('\t'),
            'b' => out.push('\u{0008}'),
            'f' => out.push('\u{000c}'),
            'u' => {
                if *at + 4 > chars.len() {
                    return None;
                }
                let hex: String = chars[*at..*at + 4].iter().collect();
                *at += 4;
                let code = u32::from_str_radix(&hex, 16).ok()?;
                out.push(char::from_u32(code)?);
            }
            _ => out.push(next),
        }
    }

    None
}

fn number(chars: &[char], at: &mut usize) -> Option<Json> {
    let start = *at;

    while *at < chars.len() && !",]}\r\n\t ".contains(chars[*at]) {
        *at += 1;
    }

    let text: String = chars[start..*at].iter().collect();

    text.parse::<f64>().ok().map(Json::Num)
}
